File: Striver/Trees/Q21.java
package Striver.Trees;

public class Q21{
    // Check if binary tree is symmetric or not

    public boolean isSymmetric(Node root) {
        if(root == null) return true;
        return areSymmetricSubtrees(root.left, root.right);
    }

    public boolean areSymmetricSubtrees(Node leftSubTree, Node rightSubTree){
        if(leftSubTree == null || rightSubTree == null) return leftSubTree == rightSubTree;
        if(leftSubTree.data != rightSubTree.data) return false;
        return areSymmetricSubtrees(leftSubTree.left, rightSubTree.right) && areSymmetricSubtrees(leftSubTree.right, rightSubTree.left);
    }

    public static void main(String[] args) {
        Q21 q = new Q21();
        Node root = new Node(1);
        root.left = new Node(2);
        root.right = new Node(2);
        System.out.println(q.isSymmetric(root));   
    }
}

File: Striver/Trees/Q21.py
# check if binary tree is symmetric or not

from Node import Node

def isSymmetric(root):
    if root is None:
        return True
    return areSymmetricSubtrees(root.left, root.right)

def areSymmetricSubtrees(leftSubtree, rightSubtree):
    if leftSubtree is None or rightSubtree is None:
        return leftSubtree == rightSubtree
    
    if leftSubtree.data != rightSubtree.data:
        return False
    
    return areSymmetricSubtrees(leftSubtree.left, rightSubtree.right) and areSymmetricSubtrees(leftSubtree.right, rightSubtree.left)

root = Node(1)
root.left = Node(2)
root.right = Node(2)
root.left.left = Node(4)
root.left.right = Node(6)
root.left.right.left = Node(8)
root.left.right.right = Node(9)
root.right.left = Node(6)
root.right.right = Node(4)
root.right.left.left = Node(9)
root.right.left.right = Node(8)

print("Symmetric" if isSymmetric(root) else "Not symmetric")